### PR TITLE
Feature/refactor stats inventory

### PR DESCRIPTION
**Pull request by qonl**
Refactor:

- adds setter and getter methods for `Character.stats`, to be able to mutate and update them in the future.
- creates new `Inventory` class to manage inventory.

### CODE REVIEW

File: dark-souls-class.js [hunk 1]
```diff
@@ -5,33 +5,64 @@ const readline = require('readline');
 const rl = readline.createInterface({
     input: process.stdin,
     output: process.stdout
-  });
+});
+
+class Inventory {
```
Comment (clinzy25) [inline dark-souls-class.js:1]: Inventory add/remove is dope

File: dark-souls-class.js [hunk 2]
```diff
@@ -5,33 +5,64 @@ const readline = require('readline');
...
-                    'intellegence', 
-                    'faith'
-                    ] 
- 
+        this.inventory = inventory;
+        this._stats = stats;
     }
+
+    // setters => change (mutate) properties, with validation
+    set stats(newStats) {
```
Comment (clinzy25) [inline dark-souls-class.js:40]: The setter doesn't seem to work. Try running the following tests

```
let guy = new Knight();
guy.level = '9';
//should throw Error

log(guy.stats)
log(guy.level)
log(guy.stats.level)
```
I tried to fix but I'm not sure what the issue is.

Comment (qonl) [reply dark-souls-class.js:40]: The setter is accessed through `guy.stats`. To level up you would do something like

`guy.stats = {
    level = 9,
}`

The `guy.stats` is also the getter, I.e. `console.log(guy.stats.level);`.

The refactor doesn't currently place each stat directly on the `Character`, but could easily be added by doing something like this in the `Knight` constructor

`for (let key in this.stats) {
    this[key] = this.stats[key];
}`

the setter doesn't currently support this though

Comment (clinzy25) [reply dark-souls-class.js:40]: `guy.stats = { level: 9 }
`
Adds stats

Why does `Character` need the stats? Won't they be different for each new character?

Comment (qonl) [reply dark-souls-class.js:40]: They are different. Take the following:

```
const warrior = new Warrior('Q', 'Male', 'Twin Humanities', 'Thin');

console.log(warrior.stats);
// Initial stats, unique to Warrior class:
// {
//     level: 4,
//     vitality: 11,
//     attunement: 8,
//     endurance: 12,
//     strength: 13,
//     dexterity: 13,
//     resistance: 11,
//     intelligence: 9,
//     faith: 9
// }

// We only need to update the stats that are necessary
const newWarriorStats = {
    level: 10,
}

// Setter, not obvious that this is a function, therefore
// hiding implementation details from the consumer
warrior.stats = newWarriorStats;

console.log(warrior.stats);
// Updated stats, only level has changed:
// {
//     level: 10,
//     vitality: 11,
//     attunement: 8,
//     endurance: 12,
//     strength: 13,
//     dexterity: 13,
//     resistance: 11,
//     intelligence: 9,
//     faith: 9
// }
```

This is different from Knight, confirmed by the following:

```
const knight = new Knight('Q', 'Male', 'Twin Humanities', 'Thin');

console.log(knight.stats);
// Initial stats, unique to Knight:
// {
//     level: 5,
//     vitality: 14,
//     attunement: 10,
//     endurance: 10,
//     strength: 11,
//     dexterity: 11,
//     resistance: 10,
//     intelligence: 9,
//     faith: 11
// }

// We only need to update the stats that are necessary
const newKnightStats = {
    level: 9,
}

// Setter, not obvious that this is a function, therefore
// hiding implementation details from the consumer
knight.stats = newKnightStats;

console.log(knight.stats);
// Updated stats, only level has changed:
// {
//     level: 9,
//     vitality: 14,
//     attunement: 10,
//     endurance: 10,
//     strength: 11,
//     dexterity: 11,
//     resistance: 10,
//     intelligence: 9,
//     faith: 11
// }
```

Each Dark Souls class is an extension of `Character`. This is the beauty of inheritance. I hope this makes it clear.

Comment (clinzy25) [reply dark-souls-class.js:40]: When you said 'the refactor doesn't currently place each stat directly on the `Character`' I thought you meant `class Character`, not the character we create with `new`. 

You're saying any `new Character` we create doesn't actually own the `stats` we are reading and modifying. Its just reading(??) them from Knight or Warrior

```
let guy = new Knight()
guy.hasOwnProperty(guy.stats)

false

//also the way it actually appears when you log(guy)

guy.hasOwnProperty(guy._stats)

false
```
But this makes my brain explode. That's prototypal inheritance. If you could add whatever is needed to the setter I'll probably get it, and merge this.

I do like the changes, I'm just trying to understand. I honestly thought I knew this shit already ......

**Pull request merged**